File: RIRImageResizing/Classes/NSObject/RIRResizeImageOperationParameters.m
//
//  RIRResizeImageOperationParameters.m
//  Pods
//
//  Created by Richard Reitzfeld on 11/21/16.
//
//

#import "RIRResizeImageOperationParameters.h"

#import <ResplendentUtilities/RUConditionalReturn.h>





@interface RIRResizeImageOperationParameters ()

#pragma mark - newSize
@property (nonatomic, assign) CGSize newSize;

#pragma mark - resizeMode
@property (nonatomic, assign) UIImage_RIRResizing_ResizeMode resizeMode;

#pragma mark - scalse
@property (nonatomic, assign) CGFloat scale;

@end





@implementation RIRResizeImageOperationParameters

#pragma mark - NSObject
-(instancetype)init
{
    kRUConditionalReturn_ReturnValueNil(YES, YES);
    return [self init_with_newSize:CGSizeZero
                        resizeMode:UIImage_RIRResizing_ResizeMode_none
                             scale:0.0f];
}

#pragma mark - init
-(instancetype)init_with_newSize:(CGSize)newSize
                      resizeMode:(UIImage_RIRResizing_ResizeMode)resizeMode
                           scale:(CGFloat)scale
{
    kRUConditionalReturn_ReturnValueNil(newSize.height <= 0 || newSize.width <= 0, YES);
    kRUConditionalReturn_ReturnValueNil(UIImage_RIRResizing_ResizeMode__isInRange(resizeMode) == NO, YES);
    
    if (self = [super init])
    {
        [self setNewSize:newSize];
        [self setResizeMode:resizeMode];
        [self setScale:scale];
    }
    
    return self;
}

@end

File: RIRImageResizing/Classes/NSObject/RIRResizeImageOperationParameters.swift
//
//  RIRResizeImageOperationParameters.swift
//  Pods-RIRImageResizing
//
//  Created by Benjamin Maer on 4/7/19.
//

import Foundation

public class RIRResizeImageOperationParameters {
    // MARK: - Constants
    public struct PublicConstants {
        public static let allowLargerResizedDimensionsDefaultValue = RIRResizeImageOperationParametersObjCLegacy.allowLargerResizedDimensionsDefaultValue
    }
    
    let newSize: CGSize
    let resizeType: RIRImageResizeType
    let scale: CGFloat?
    
    let allowLargerResizedDimensions: Bool
    
    // MARK: - Init
    enum InitError: Error {
        case badNewSizeWidth, badNewSizeHeight
    }
    
    public init(newSize: CGSize,
                resizeType: RIRImageResizeType,
                scale: CGFloat? = 0,
                allowLargerResizedDimensions: Bool = PublicConstants.allowLargerResizedDimensionsDefaultValue) throws {
        guard newSize.width > 0 else    { throw InitError.badNewSizeWidth }
        guard newSize.height > 0 else   { throw InitError.badNewSizeHeight }
        
        self.newSize = newSize
        self.resizeType = resizeType
        self.scale = scale
        self.allowLargerResizedDimensions = allowLargerResizedDimensions
    }
}

extension RIRResizeImageOperationParameters: CustomStringConvertible {
    public var description: String {
        return [
            "newSize: \(newSize)",
            "resizeType: \(resizeType)",
            "scale: \(String(describing: scale))",
        ].joined(separator: "\n")
    }
}

@objc(RIRResizeImageOperationParameters) public class RIRResizeImageOperationParametersObjCLegacy: NSObject {
    // MARK: - Constants
    @objc public static let allowLargerResizedDimensionsDefaultValue = false
    
    @objc var newSize: CGSize { return swiftInstance.newSize }
    @objc let resizeType: RIRImageResizeTypeObjCLegacy
    @objc var scale: CGFloat { return swiftInstance.scale ?? 0 }
    @objc var allowLargerResizedDimensions: Bool { return swiftInstance.allowLargerResizedDimensions }
    
    // MARK: - Swift Instance
    let swiftInstance: RIRResizeImageOperationParameters
    
    // MARK: - Init
    @objc(init_with_newSize:resizeMode:scale:allowLargerResizedDimensions:) public init?(newSize: CGSize, resizeType: RIRImageResizeTypeObjCLegacy, scale: CGFloat = 0, allowLargerResizedDimensions: Bool) {
        self.resizeType = resizeType
        do {
            try swiftInstance = RIRResizeImageOperationParameters(newSize: newSize, resizeType: resizeType.swiftValue, scale: scale, allowLargerResizedDimensions: allowLargerResizedDimensions)
        } catch {
            assertionFailure("Failed to create `swiftParameters`")
            return nil
        }
    }
    
    @objc(init_with_newSize:resizeMode:scale:) public convenience init?(newSize: CGSize, resizeType: RIRImageResizeTypeObjCLegacy, scale: CGFloat = 0) {
        self.init(newSize: newSize, resizeType: resizeType, scale: scale, allowLargerResizedDimensions: RIRResizeImageOperationParametersObjCLegacy.allowLargerResizedDimensionsDefaultValue)
    }
}
